# Stop words tutorial


In [62]:
import spacy

from spacy.lang.en.stop_words import STOP_WORDS

In [63]:
print(STOP_WORDS)

{'nothing', 'by', 'an', 'before', 'few', 'although', 'both', 'whom', 'fifteen', 'hereby', 'either', 'serious', 'give', 'nowhere', 'what', 'ours', 'hereupon', '’s', 'until', 'it', 'whether', 'well', 'top', '‘m', 'would', 'whereupon', '’m', 'whereby', 'towards', 'only', 'many', 'why', 'around', 'noone', 'meanwhile', 'throughout', 'was', 'n’t', 'name', 'yourself', 'among', 'hereafter', 'every', 'whither', 'somehow', 'below', 'forty', 'eight', 'whose', 'across', 'be', 'to', 'upon', 'everyone', 'this', 'when', 'show', 'for', 'have', 'back', 're', 'wherein', 'whole', 'using', "'s", 'another', 'really', 'nine', 'sixty', '‘s', 'could', 'if', 'must', 'regarding', '’ll', 'amongst', 'seems', '‘re', 'mine', 'my', 'them', 'never', 'are', 'see', 'then', 'next', 'is', 'empty', 'each', 'those', 'cannot', '’ve', 'will', 'make', 'here', 'nevertheless', 'in', 'became', 'along', 'because', 'indeed', "'re", 'seeming', 'anyway', 'these', 'even', 'others', 'several', 'himself', 'herself', 'since', 'third', '

In [64]:
len(STOP_WORDS)

326

In [65]:
nlp = spacy.load("en_core_web_sm")

doc = nlp("We just opened our wings, the flying part is coming soon")

for token in doc:
    if token.is_stop:
        print(token)

We
just
our
the
part
is


In [66]:
def preprocess(text):
    doc = nlp(text)
    no_stop_words = [token.text for token in doc if not token.is_stop]
    return " ".join(no_stop_words)       


# Also the above is same as 
'''
def preprocess1(text):
    doc = nlp(text)
    no_stop_words = []
    for token in doc:
        if not token.is_stop:
            no_stop_words.append(token.text)
    return " ".join(no_stop_words)
'''

'\ndef preprocess1(text):\n    doc = nlp(text)\n    no_stop_words = []\n    for token in doc:\n        if not token.is_stop:\n            no_stop_words.append(token.text)\n    return " ".join(no_stop_words)\n'

In [67]:
preprocess("Musk wants time to prepare for a trial over his")

'Musk wants time prepare trial'

In [68]:
preprocess("The other is not other but your divine brother")

'divine brother'

<b> Remove stop words from pandas dataframe text column </b>

Dataset is downloaded from: https://www.kaggle.com/datasets/jbencina/department-of-justice-20092018-press-releases It contains press releases of different court cases from depart of justice (DOJ). The releases contain information such as outcomes of criminal cases, notable actions taken against felons, or other updates about the current administration.

In [69]:
import pandas as pd

df = pd.read_json("doj_press.json",lines=True)

df.shape

(13087, 6)

In [70]:
df.head()

id                                              title  \
0     None       Convicted Bomb Plotter Sentenced to 30 Years   
1  12-919   $1 Million in Restitution Payments Announced t...   
2  11-1002  $1 Million Settlement Reached for Natural Reso...   
3   10-015  10 Las Vegas Men Indicted \r\nfor Falsifying V...   
4   18-898  $100 Million Settlement Will Speed Cleanup Wor...   

                                            contents  \
0  PORTLAND, Oregon. – Mohamed Osman Mohamud, 23,...   
1    WASHINGTON – North Carolina’s Waccamaw River...   
2        BOSTON– A $1-million settlement has been...   
3    WASHINGTON—A federal grand jury in Las Vegas...   
4  The U.S. Department of Justice, the U.S. Envir...   

                        date         topics  \
0  2014-10-01T00:00:00-04:00             []   
1  2012-07-25T00:00:00-04:00             []   
2  2011-08-03T00:00:00-04:00             []   
3  2010-01-08T00:00:00-05:00             []   
4  2018-07-09T00:00:00-04:00  [Environment]   

                                     components  
0            [National Security Division (NSD)]  
1  [Environment and Natural Resources Division]  
2  [Environment and Natural Resources Division]  
3  [Environment and Natural Resources Division]  
4  [Environment and Natural Resources Division]

In [71]:
df = df[df['topics'].str.len() != 0]
df.head()

id                                              title  \
4    18-898  $100 Million Settlement Will Speed Cleanup Wor...   
7   14-1412  14 Indicted in Connection with New England Com...   
19  17-1419  2017 Southeast Regional Animal Cruelty Prosecu...   
22  15-1562  21st Century Oncology to Pay $19.75 Million to...   
23  17-1404  21st Century Oncology to Pay $26 Million to Se...   

                                             contents  \
4   The U.S. Department of Justice, the U.S. Envir...   
7   A 131-count criminal indictment was unsealed t...   
19  The United States Attorney’s Office for the Mi...   
22  21st Century Oncology LLC, has agreed to pay $...   
23  21st Century Oncology Inc. and certain of its ...   

                         date                                 topics  \
4   2018-07-09T00:00:00-04:00                          [Environment]   
7   2014-12-17T00:00:00-05:00                  [Consumer Protection]   
19  2017-12-14T00:00:00-05:00                          [Environment]   
22  2015-12-18T00:00:00-05:00  [False Claims Act, Health Care Fraud]   
23  2017-12-12T00:00:00-05:00  [Health Care Fraud, False Claims Act]   

                                           components  
4        [Environment and Natural Resources Division]  
7                                    [Civil Division]  
19  [Environment and Natural Resources Division, U...  
22                                   [Civil Division]  
23           [Civil Division, USAO - Florida, Middle]

In [72]:
df.shape

(4688, 6)

<b> Here we have removed all the data which did not contained the  topics in the dataset </b>

<b> i.e from 13807 to 4688 </b>

In [73]:
df = df.head(100)
df.head()

id                                              title  \
4    18-898  $100 Million Settlement Will Speed Cleanup Wor...   
7   14-1412  14 Indicted in Connection with New England Com...   
19  17-1419  2017 Southeast Regional Animal Cruelty Prosecu...   
22  15-1562  21st Century Oncology to Pay $19.75 Million to...   
23  17-1404  21st Century Oncology to Pay $26 Million to Se...   

                                             contents  \
4   The U.S. Department of Justice, the U.S. Envir...   
7   A 131-count criminal indictment was unsealed t...   
19  The United States Attorney’s Office for the Mi...   
22  21st Century Oncology LLC, has agreed to pay $...   
23  21st Century Oncology Inc. and certain of its ...   

                         date                                 topics  \
4   2018-07-09T00:00:00-04:00                          [Environment]   
7   2014-12-17T00:00:00-05:00                  [Consumer Protection]   
19  2017-12-14T00:00:00-05:00                          [Environment]   
22  2015-12-18T00:00:00-05:00  [False Claims Act, Health Care Fraud]   
23  2017-12-12T00:00:00-05:00  [Health Care Fraud, False Claims Act]   

                                           components  
4        [Environment and Natural Resources Division]  
7                                    [Civil Division]  
19  [Environment and Natural Resources Division, U...  
22                                   [Civil Division]  
23           [Civil Division, USAO - Florida, Middle]

In [74]:
df.shape

(100, 6)

In [75]:
df["contents_new"] = df.contents.apply(preprocess)

In [76]:
df

id                                              title  \
4     18-898  $100 Million Settlement Will Speed Cleanup Wor...   
7    14-1412  14 Indicted in Connection with New England Com...   
19   17-1419  2017 Southeast Regional Animal Cruelty Prosecu...   
22   15-1562  21st Century Oncology to Pay $19.75 Million to...   
23   17-1404  21st Century Oncology to Pay $26 Million to Se...   
..       ...                                                ...   
316  15-1359  Alaska Plastic Surgeon Convicted of Wire Fraud...   
318   16-396  Alaska Plastic Surgeon Sentenced to Prison for...   
321   17-736  Alaskan Commercial Fishing Couple Charged with...   
322   18-717  Alaskan Husband And Wife Plead Guilty To Willf...   
324  16-1345        Alaskan Oncologist Indicted for Tax Evasion   

                                              contents  \
4    The U.S. Department of Justice, the U.S. Envir...   
7    A 131-count criminal indictment was unsealed t...   
19   The United States Attorney’s Office for the Mi...   
22   21st Century Oncology LLC, has agreed to pay $...   
23   21st Century Oncology Inc. and certain of its ...   
..                                                 ...   
316  Doctor Hid Millions in Secret Accounts in Pana...   
318  Defendant Concealed Bank Accounts in Panama an...   
321  An Alaskan couple was charged in federal court...   
322  A husband and wife pleaded guilty yesterday to...   
324  A resident of Big Lake, Alaska was indicted on...   

                          date                                 topics  \
4    2018-07-09T00:00:00-04:00                          [Environment]   
7    2014-12-17T00:00:00-05:00                  [Consumer Protection]   
19   2017-12-14T00:00:00-05:00                          [Environment]   
22   2015-12-18T00:00:00-05:00  [False Claims Act, Health Care Fraud]   
23   2017-12-12T00:00:00-05:00  [Health Care Fraud, False Claims Act]   
..                         ...                                    ...   
316  2015-11-04T00:00:00-05:00                                  [Tax]   
318  2016-04-04T00:00:00-04:00                                  [Tax]   
321  2017-07-26T00:00:00-04:00                                  [Tax]   
322  2018-06-01T00:00:00-04:00                                  [Tax]   
324  2016-11-17T00:00:00-05:00                                  [Tax]   

                                            components  \
4         [Environment and Natural Resources Division]   
7                                     [Civil Division]   
19   [Environment and Natural Resources Division, U...   
22                                    [Civil Division]   
23            [Civil Division, USAO - Florida, Middle]   
..                                                 ...   
316                                     [Tax Division]   
318                                     [Tax Division]   
321                      [Tax Division, USAO - Alaska]   
322                                     [Tax Division]   
324                                     [Tax Division]   

                                          contents_new  
4    U.S. Department Justice , U.S. Environmental P...  
7    131 - count criminal indictment unsealed today...  
19   United States Attorney Office Middle District ...  
22   21st Century Oncology LLC , agreed pay $ 19.75...  
23   21st Century Oncology Inc. certain subsidiarie...  
..                                                 ...  
316  Doctor Hid Millions Secret Accounts Panama Ala...  
318  Defendant Concealed Bank Accounts Panama Costa...  
321  Alaskan couple charged federal court Juneau , ...  
322  husband wife pleaded guilty yesterday counts w...  
324  resident Big Lake , Alaska indicted counts tax...  

[100 rows x 7 columns]

In [77]:
len(df.contents[4])

6286

In [78]:
len(df.contents_new[4])

4810

In [79]:
df.contents[4][:300]

'The U.S. Department of Justice, the U.S. Environmental Protection Agency (EPA), and the Rhode Island Department of Environmental Management (RIDEM) announced today that two subsidiaries of Stanley Black & Decker Inc.—Emhart Industries Inc. and Black & Decker Inc.—have agreed to clean up dioxin conta'

In [80]:
df.contents_new[4][:300]

'U.S. Department Justice , U.S. Environmental Protection Agency ( EPA ) , Rhode Island Department Environmental Management ( RIDEM ) announced today subsidiaries Stanley Black & Decker Inc.—Emhart Industries Inc. Black & Decker Inc.—have agreed clean dioxin contaminated sediment soil Centredale Manor'

In [81]:
print((len(df.contents_new[4])/len(df.contents[4]))*100)

76.51924912503976


Here we can see that the new content is 24 less then the original content 

# Examples where removing stop words can create a problem


<b> (1) Sentiment detection: Not always but in some cases, based on your dataset it can change the sentiment of a sentence if you remove stop words </b>

In [82]:
preprocess("this is a good movie")

'good movie'

In [83]:
preprocess("this is not a good movie")

'good movie'

<b> 
(2) Language translation: Say you want to translate following sentence from english to telugu. Before actual translation if you remove stop words and then translate, it will produce horrible result</b>

In [84]:
preprocess("how are you doing Nami?")

'Nami ?'

<b> (3) Chat bot or any Q&A system </b>

In [85]:
preprocess("I don't find yoga mat on your website. Can you help?")

'find yoga mat website . help ?'